GAE/src/org/waterforpeople/mapping/app/web/CronCommanderServlet.java

```diff
@@ -68,21 +74,56 @@ public void doGet(HttpServletRequest req, HttpServletResponse resp)
             generateNotifications();
         } else if ("purgeDeviceFileJobQueueRecords".equals(action)) {
             purgeDeviceFileJobQueueRecords();
+        } else if ("purgeExpiredDevices".equals(action)) {
+            purgeExpiredDevices();
         } else if ("purgeReportRecords".equals(action)) {
             purgeReportRecords();
         }
     }
 
     /**
-     * scans for and deletes Report entries that are either more than one year old
+     * scans for and deletes Device entries that have not been seen in more than two years
+     */
+    private void purgeExpiredDevices() {
+        Calendar deadline = Calendar.getInstance();
+        deadline.add(Calendar.YEAR, TWO_YEARS_AGO);
+        log.info("Starting scan for Devices not seen since: " + deadline.getTime());
+        DeviceDAO deviceDao = new DeviceDAO();
+        DeviceSurveyJobQueueDAO dsjqDao = new DeviceSurveyJobQueueDAO();
+        DeviceFileJobQueueDAO dfjqDao = new DeviceFileJobQueueDAO();
+        SurveyAssignmentDAO saDao = new SurveyAssignmentDAO();
+        List<Device> deviceList = deviceDao.listAllWithBeaconBefore(deadline.getTime());
+        log.info("Found " + deviceList.size() + " old Devices");
+
+        for (Device d: deviceList) { //Clean up everything referencing this device
+        	
+        	List<DeviceSurveyJobQueue> djql = dsjqDao.get(d.getPhoneNumber(), d.getEsn(), d.getAndroidId());
+        	if (djql.size() > 0) {
+        		log.fine("Deleting " + djql.size() + " form assignments for device " + d.getKey().getId());
+        		dsjqDao.delete(djql);
+        	}
+        	List<DeviceFileJobQueue> dfql = dfjqDao.listByDeviceId(d.getKey().getId());
```

Comment:
Maybe you should extract d.getKey().getId() to var

Reply:
We do it this way in lots of places. It is not expensive...